### PR TITLE
gyp: Export LD_LIBRARY_PATH quoted in case $(builddir) contains spaces

### PR DESCRIPTION
**Pull request by melloc**
##### Checklist

- [x] `npm install && npm test` passes
- [x] tests are included 
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#commit-message-guidelines)

##### Description of change

This change will make sure that the path to the build directory gets properly quoted and escaped when running actions.

### CODE REVIEW

**Comment by refack** [comment]
Hello @melloc, thank you for the contribution.
This LGTM (apart from a few style nits), and I'm running `node-gyp`'s CI on this: https://ci.nodejs.org/view/All/job/nodegyp-test-pull-request/34/
But what I would ask you to do, if you are so inclined, is it to submit this upstream to: https://gyp.gsrc.io/docs/Hacking.md
(If you rather not deal with the hassle, please let me know, and I'll submit it on your behalf.)

File: gyp/pylib/gyp/generator/make.py [hunk 1]
```diff
@@ -916,11 +916,12 @@ def WriteActions(self, actions, extra_sources, extra_outputs,
       # libraries, but until everything is made cross-compile safe, also use
       # target libraries.
       # TODO(piman): when everything is cross-compile safe, remove lib.target
-      self.WriteLn('cmd_%s = LD_LIBRARY_PATH=$(builddir)/lib.host:'
-                   '$(builddir)/lib.target:$$LD_LIBRARY_PATH; '
+      subst_builddir = '$(subst \',\'\\\'\',$(builddir))'
```
Comment (refack) [inline gyp/pylib/gyp/generator/make.py]: nit: I'd rather you use the `"` string delimiter and the `r` prefix so you get:
```python
     subst_builddir = r"$(subst ', '\'', $(builddir))"
```
Also IMHO the arg to `self.WriteLn` should be constructed before the call
```
     action_cmd = "..." \
       "..." % (...)
     self.WriteLn(action_cmd)
```

**Comment by refack** [comment]
Seems like the new test fails on Windows:
```tap
module.js:487
    throw err;
    ^

Error: Cannot find module 'name's (weird)'
    at Function.Module._resolveFilename (module.js:485:15)
    at Function.Module._load (module.js:437:25)
    at Module.require (module.js:513:17)
    at require (internal/module.js:11:18)
    at [eval]:1:1
    at ContextifyScript.Script.runInThisContext (vm.js:44:33)
    at Object.runInThisContext (vm.js:116:38)
    at Object.<anonymous> ([eval]-wrapper:6:22)
    at Module._compile (module.js:569:30)
    at evalScript (bootstrap_node.js:432:27)
child_process.js:591
    throw err;
    ^

Error: Command failed: node -e require('name\'s (weird)')
module.js:487
    throw err;
    ^

Error: Cannot find module 'name's (weird)'
    at Function.Module._resolveFilename (module.js:485:15)
    at Function.Module._load (module.js:437:25)
    at Module.require (module.js:513:17)
    at require (internal/module.js:11:18)
    at [eval]:1:1
    at ContextifyScript.Script.runInThisContext (vm.js:44:33)
    at Object.runInThisContext (vm.js:116:38)
    at Object.<anonymous> ([eval]-wrapper:6:22)
    at Module._compile (module.js:569:30)
    at evalScript (bootstrap_node.js:432:27)


    at checkExecSyncError (child_process.js:568:13)
    at execFileSync (child_process.js:588:13)
    at runSpacedName (c:\workspace\nodegyp-test-commit\nodes\win2012r2\test\test-actions.js:15:10)
    at c:\workspace\nodegyp-test-commit\nodes\win2012r2\test\test-actions.js:29:19
    at ChildProcess.exithandler (child_process.js:244:7)
    at emitTwo (events.js:125:13)
    at ChildProcess.emit (events.js:213:7)
    at maybeClose (internal/child_process.js:887:16)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:208:5)

File: test/test-actions.js [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+var path = require('path')
+
+var execFileSync = child_process.execFileSync
+var execFile = child_process.execFile
+
+var addonPath = path.resolve(__dirname, 'node_modules', 'name\'s (weird)')
+var nodeGyp = path.resolve(__dirname, '..', 'bin', 'node-gyp.js')
+
+function runSpacedName() {
+  var testCode = "require('name\\'s (weird)')"
```
Comment (refack) [inline test/test-actions.js]: You can't put a `\` in a Windows path. But you can put an unescaped `'`.
You could try using `require("${addonPath}")` but YMMV

File: test/test-actions.js [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+  // Set the loglevel otherwise the output disappears when run via 'npm test'
+  var cmd = [ nodeGyp, 'rebuild', '-C', addonPath, '--loglevel=verbose' ]
+
+  var proc = execFile(process.execPath, cmd, function (err, stdout, stderr) {
+    var logLines = stderr.toString().trim().split(/\r?\n/)
+    var lastLine = logLines[logLines.length-1]
+    t.strictEqual(err, null)
+    t.strictEqual(lastLine, 'gyp info ok', 'should end in ok')
+    t.strictEqual(runSpacedName().trim(), 'world')
```
Comment (refack) [inline test/test-actions.js]: You should the `.trim()` into `runSpacedName`

File: test/test-actions.js [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+}
+
+test('build addon w/ actions and spaces in path', function (t) {
+  t.plan(3)
+
+  // Set the loglevel otherwise the output disappears when run via 'npm test'
+  var cmd = [ nodeGyp, 'rebuild', '-C', addonPath, '--loglevel=verbose' ]
+
+  var proc = execFile(process.execPath, cmd, function (err, stdout, stderr) {
+    var logLines = stderr.toString().trim().split(/\r?\n/)
```
Comment (refack) [inline test/test-actions.js]:
```js
  const lastLine = stderr.toString().trim().split(path.EOL).pop();
```

File: test/test-actions.js (hunk 3 above)
Comment (refack) [inline test/test-actions.js]: You should put a comment explaining where "world" comes from.

**Comment by melloc** [comment]
@refack Thanks for the feedback. I suspect the Windows tests are failing because the shell script is failing and not creating `hello.js`, although I'm not sure why this isn't causing the `node-gyp` to exit non-zero. I'll modify it to run a batch script on Windows, and the shell script on everything else, and see if I can find a Windows environment to test it in.

Once I have that I'll look into what it would take to get it submitted upstream, too.

**Comment by cup** [comment]
https://github.com/nodejs/node-gyp/issues/65

**Comment by rvagg** [comment]
not sure if this is still needed, if so and it belongs here please open a fresh PR for it, otherwise maybe https://github.com/refack/GYP is the right place for it

**Pull request closed without merging**